File: ducklin-core/src/main/java/io/github/ducklin/migration/researches/Research.java
package io.github.ducklin.migration.researches;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;
import java.util.function.Consumer;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;

import io.github.ducklin.api.items.DuckItem;
import io.github.ducklin.core.Slimefun;
import org.apache.commons.lang3.Validate;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.GameMode;
import org.bukkit.Keyed;
import org.bukkit.NamespacedKey;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

import io.github.ducklin.migration.events.PlayerPreResearchEvent;
import io.github.ducklin.migration.items.ItemGroup;
import io.github.ducklin.migration.items.ItemState;
import io.github.ducklin.migration.player.PlayerProfile;


public class Research implements Keyed {

    private final NamespacedKey key;
    private final int id;
    private final String name;
    private boolean enabled = true;
    private int cost;

    private final List<DuckItem> items = new LinkedList<>();

    public Research(@Nonnull NamespacedKey key, int id, @Nonnull String defaultName, int defaultCost) {
        Validate.notNull(key, "A NamespacedKey must be provided");
        Validate.notNull(defaultName, "A default name must be specified");

        this.key = key;
        this.id = id;
        this.name = defaultName;
        this.cost = defaultCost;
    }

    @Override
    public @Nonnull NamespacedKey getKey() {
        return key;
    }

    public boolean isEnabled() {
        return Slimefun.getRegistry().isResearchingEnabled() && enabled;
    }

    @Deprecated
    public int getID() {
        return id;
    }


    public @Nonnull String getName(@Nonnull Player p) {
        String localized = Slimefun.getLocalization().getResearchName(p, key);
        return localized != null ? localized : name;
    }

    public @Nonnull String getUnlocalizedName() {
        return ChatColor.stripColor(name);
    }

    public int getCost() {
        return cost;
    }

    public void setCost(int cost) {
        if (cost < 0) {
            throw new IllegalArgumentException("Research cost must be zero or greater!");
        }

        this.cost = cost;
    }


    public void addItems(DuckItem... items) {
        for (DuckItem item : items) {
            if (item != null) {
                item.setResearch(this);
            }
        }
    }

    @Nonnull
    public Research addItems(ItemStack... items) {
        for (ItemStack item : items) {
            DuckItem sfItem = DuckItem.getByItem(item);

            if (sfItem != null) {
                sfItem.setResearch(this);
            }
        }

        return this;
    }

    @Nonnull
    public List<DuckItem> getAffectedItems() {
        return items;
    }

    public boolean hasEnabledItems() {
        for (DuckItem item : items) {
            if (item.getState() == ItemState.ENABLED) {
                return true;
            }
        }
        return false;
    }


    @ParametersAreNonnullByDefault
    public void unlockFromGuide(SlimefunGuideImplementation guide, Player player, PlayerProfile profile, DuckItem sfItem, ItemGroup itemGroup, int page) {
        if (!Slimefun.getRegistry().getCurrentlyResearchingPlayers().contains(player.getUniqueId())) {
            if (profile.hasUnlocked(this)) {
                guide.openItemGroup(profile, itemGroup, page);
            } else {
                PlayerPreResearchEvent event = new PlayerPreResearchEvent(player, this, sfItem);
                Bukkit.getPluginManager().callEvent(event);

                if (!event.isCancelled()) {
                    if (this.canUnlock(player)) {
                        guide.unlockItem(player, sfItem, pl -> guide.openItemGroup(profile, itemGroup, page));
                    } else {
                        Slimefun.getLocalization().sendMessage(player, "messages.not-enough-xp", true);
                    }
                }
            }
        }
    }

    public boolean canUnlock(@Nonnull Player p) {
        if (!isEnabled()) {
            return true;
        }

        boolean creativeResearch = p.getGameMode() == GameMode.CREATIVE && Slimefun.getRegistry().isFreeCreativeResearchingEnabled();
        return creativeResearch || p.getLevel() >= cost;
    }


    public void unlock(@Nonnull Player p, boolean instant) {
        unlock(p, instant, null);
    }


    public void unlock(@Nonnull Player p, boolean isInstant, @Nullable Consumer<Player> callback) {
        PlayerProfile.get(p, new PlayerResearchTask(this, isInstant, callback));
    }

    public void register() {
        Slimefun.getResearchCfg().setDefaultValue("enable-researching", true);
        String path = key.getNamespace() + '.' + key.getKey();

        if (Slimefun.getResearchCfg().contains(path + ".enabled") && !Slimefun.getResearchCfg().getBoolean(path + ".enabled")) {
            for (DuckItem item : new ArrayList<>(items)) {
                if (item != null) {
                    item.setResearch(null);
                }
            }

            enabled = false;
            return;
        }

        Slimefun.getResearchCfg().setDefaultValue(path + ".cost", getCost());
        Slimefun.getResearchCfg().setDefaultValue(path + ".enabled", true);

        setCost(Slimefun.getResearchCfg().getInt(path + ".cost"));
        enabled = true;

        Slimefun.getRegistry().getResearches().add(this);
    }

    @Nonnull
    public static Optional<Research> getResearch(@Nullable NamespacedKey key) {
        if (key == null) {
            return Optional.empty();
        }

        for (Research research : Slimefun.getRegistry().getResearches()) {
            if (research.getKey().equals(key)) {
                return Optional.of(research);
            }
        }

        return Optional.empty();
    }

    @Override
    public String toString() {
        return "Research (" + getKey() + ')';
    }
}

File: ducklin-api/src/main/kotlin/io/github/ducklin/api/researches/Research.kt
package io.github.ducklin.api.researches

import io.github.ducklin.api.DucklinAPI
import io.github.ducklin.api.items.DucklinItem
import io.github.ducklin.api.items.ItemState
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.Keyed
import org.bukkit.NamespacedKey
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

class Research(
    private val key: NamespacedKey,
    val id: Int,
    private val defaultName: String,
    cost: Int
) : Keyed {

    var isEnabled: Boolean = true; private set

    var cost: Int = cost
        set(value) {
            require(value > 0) {"Research cost must be zero or greater!"}
            field = value
        }

    private val items: MutableList<DucklinItem> = mutableListOf()

    override fun getKey(): NamespacedKey = key

    val isResearchingEnabled: Boolean
        get() = DucklinAPI.registry.enableResearches && isEnabled

    @Deprecated("Use key instead")
    fun getID(): Int = id

    fun getName(player: Player): Component =
        MiniMessage.miniMessage().deserialize(
            DucklinAPI.services.localizationService.getResearchName(player, key) ?: defaultName
        )

    fun getUnlocalizedName(): Component =MiniMessage.miniMessage().deserialize(defaultName)

    fun addItems(vararg items: DucklinItem?) {
        items.filterNotNull().forEach { it.research = this }
    }

    fun addItems(vararg stacks: ItemStack?): Research = apply {
        stacks.forEach { stack ->
            DucklinItem.getByItem(stack)?.let { it.research = this }
        }
    }

    val affectedItems: List<DucklinItem> get() = items.toList()

    fun hasEnabledItems(): Boolean = items.any { it.state == ItemState.ENABLED }

    fun unlockFromGuide(
        guide: SlimefunGuideImplementation,
        player: Player,
        profile: PlayerProfile,
        sfItem: DucklinItem,
        itemGroup: ItemGroup,
        page: Int
    ) {
        if (player.uniqueId in Slimefun.getRegistry().currentlyResearchingPlayers) return

        if (profile.hasUnlocked(this)) {
            guide.openItemGroup(profile, itemGroup, page)
            return
        }

        val event = PlayerPreResearchEvent(player, this, sfItem)
        Bukkit.getPluginManager().callEvent(event)
        if (event.isCancelled) return

        if (canUnlock(player)) {
            guide.unlockItem(player, sfItem) { guide.openItemGroup(profile, itemGroup, page) }
        } else {
            Slimefun.getLocalization().sendMessage(player, "messages.not-enough-xp", true)
        }
    }

    fun canUnlock(player: Player): Boolean {
        if (!isResearchingEnabled) return true
        val creativeFree = player.gameMode == GameMode.CREATIVE &&
                Slimefun.getRegistry().isFreeCreativeResearchingEnabled
        return creativeFree || player.level >= cost
    }

    fun unlock(player: Player, instant: Boolean, callback: Consumer<Player>? = null) {
        PlayerProfile.get(player, PlayerResearchTask(this, instant, callback))
    }

    fun register() {
        val cfg = Slimefun.getResearchCfg()
        cfg.setDefaultValue("enable-researching", true)
        val path = "${key.namespace}.${key.key}"

        if (cfg.getBoolean("$path.enabled") == false) {
            items.toList().forEach { it.research = null }
            isEnabled = false
            return
        }

        cfg.setDefaultValue("$path.cost", cost)
        cfg.setDefaultValue("$path.enabled", true)

        cost = cfg.getInt("$path.cost")
        isEnabled = true
        Slimefun.getRegistry().researches += this
    }

    companion object {
        @JvmStatic
        fun getResearch(key: NamespacedKey?): Research? =
            key?.let { k -> Slimefun.getRegistry().researches.firstOrNull { it.key == k } }
    }

    override fun toString(): String = "Research ($key)"
}

}